### PR TITLE
The FAMNM Attendance System!

### PR DESCRIPTION
**Pull request by eashwar**
# Checklist

  * [X] This update has been tested thoroughly enough to ensure there are no
        functional bugs or obvious layout issues.
  * [X] This update has been tested in both desktop and mobile layouts.
  * [X] This update does not break any existing functionality, layout, or style
        rules, unless the update is explicitly intended to remove or update such
        functionality, layout, or style rules.
  * [X] This update is free of any code intended only for debugging or testing
        purposes.
  * [X] This update follows the FAMNM Website Code Style Guide.
  * [X] This update meets the visual style requirements as specified in the
        FAMNM Website Visual Style Guide, and any additional visual elements not
        previously specified in the FAMNM Website Visual Style Guide have been
        styled so as not to clash visually with the rest of the site.

# Description

This is it. The Big Cheese. FAMNM's very own attendance tracking system!

This adds two pages to the FAMNM website:

`/attendance/`, and 
`/attendance/view/`.

# Additions

## `/attendance`

The attendance form
A handy dandy confirmation screen that tells you about issues with your submit

This is the 'form' that a meeting leader submits during their meeting to confirm who came to that meeting. It features a confirmation screen that tells you if there are any issues with your submission, *except for misspelled uniqnames*. 

## `/attendance/view`

You can view attendance of individual members
You can also view all the meetings recorded in the database!

This is the page to export all of the database's data as a CSV file; get the number of active members; view individual attendance data, and all of the meetings.

# Future Plans

We're gonna slowly add more features to this, such as:

- making the attendance viewer nicer with more views/abilities to sort
- some type of security/authentication

and other things!

### CODE REVIEW

File: _includes/title-styles.html [hunk 1]
```diff
@@ -11,6 +11,7 @@
 <link rel="shortcut icon" href="{{site.url}}/favicon.ico">
 <link rel="icon" sizes="192x192" href="{{site.url}}/img/brain_favicon_large.png">
 <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
+
```
Comment (bharadwaj-a) [inline _includes/title-styles.html]: {minor}Not sure why this is here.

File: attendance/index.html [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+            <label for='meetingType'>What type of meeting?</label>
+            <select class='form-control' id='meetingType'>
+                <option>
+                    ------
+                </option>
+            </select>
+        </div>
+        <div class="my-5">
+            <label for='meetingDate'>Today's Date</label>
+            <input type='date' id='meetingDate'>
```
Comment (bharadwaj-a) [inline attendance/index.html]: {nit} end the tag with "/>"

File: attendance/index.html [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+        </div>
+
+        <div class="my-5">
+            <label for='uniqname'>Enter uniqnames of people at meeting. Hit enter to add person.</label>
+            <br/>
+            <input type='text' id='uniqname'>
+            <br/>
+            <div id='uniqname-buttons'>
+
+            </div>
```
Comment (bharadwaj-a) [inline attendance/index.html]: {nit} does this need to be separated out?

File: attendance/index.html [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    <div class="modal-dialog" role="document">
+        <div class="modal-content">
+        <div class="modal-header">
+            <h5 class="modal-title" id="confirmMeetingTitle">Confirm Submission</h5>
+            <button type="button" class="close" data-dismiss="modal" aria-label="Close">
+            <span aria-hidden="true">&times;</span>
+            </button>
+        </div>
+        <div class="modal-body">
+            <div id="modalError" style="color:red"></div>
```
Comment (bharadwaj-a) [inline attendance/index.html]: {nit} combine to close "<.../>

File: attendance/index.html [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+            <h2>Meeting Details</h2>
+            <p id="modalMeetingType">Meeting Type: </p>
+            <p id="modalStartTime">Start Time: </p>
+            <p id="modalEndTime">End Time: </p>
+            <p id="modalDescription">Description: </p>
+            <h2>Meeting Attendees</h2>
+            <div id="modalAttendees">
+
+            </div>
```
Comment (bharadwaj-a) [inline attendance/index.html]: {nit} does this need to be separated out?

File: attendance/index.html [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+        </div>
+        <div class="modal-footer">
+            <button type="button" class="btn btn-secondary" id="closeModal" data-dismiss="modal">Close</button>
+            <button type="button" class="btn btn-primary submit-meeting" id="submit" data-dismiss="modal">Create Meeting</button>
+        </div>
```
Comment (bharadwaj-a) [inline attendance/index.html]: {nit} Indentation is off on this section

File: attendance/index.html [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                return time.slice(0, 5);
+            } else {
+                let hour = parseInt(time.slice(0, 3));
+                return `${(hour + 12)}:${time.slice(2, 4)}`;
+            }
+        }
+
+        let validateTime = (start, end) =>{
+            return parseInt(start.replace(':', '')) < parseInt(end.replace(':', ''));
+        }
```
Comment (bharadwaj-a) [inline attendance/index.html]: Do we have a preference on ```let``` vs ```const``` on functions? 
These two are one way but within the document ready function everything is ```const```

File: attendance/index.html [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+        } 
+
+        $(() => {
+            fetch('http://backend.famnm.club/meeting_type')
+            .then( (response) =>{
+                if (!response.ok) throw Error(response.statusText);
+                return response.json();
+            }).then( (data) => {
+                data.map(type => {
+                    let meetingType = $('<option></option>');
```
Comment (bharadwaj-a) [inline attendance/index.html]: {nit} switch for ```<option/>```

File: attendance/index.html [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+                startTime: '10:00',
+                defaultTime: '10:00',
+                dropdown: true,
+                interval: 30
+            });
+            
+            $('#uniqname').keypress((e) => {
+                if (e.keyCode == 13)
+                {
+                    let newButton = $('<button></button>');
```
Comment (bharadwaj-a) [inline attendance/index.html]: {nit} replace with ```<button/>```

File: attendance/index.html [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+                    newButton.attr('type', 'button');
+                    newButton.addClass('btn');
+                    newButton.addClass('btn-outline-secondary');
+                    newButton.addClass('mx-3');
+                    newButton.addClass('my-3');
+                    newButton.addClass('uniqname-btn');
```
Comment (bharadwaj-a) [inline attendance/index.html]: {nit} chain these together
```suggestion
                    newButton.addClass('btn').addClass('btn-outline-secondary').addClass('mx-3').addClass('my-3').addClass('uniqname-btn');
```
spacing is a bit off in suggestion

File: attendance/index.html [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+                uniqname_buttons.toArray().forEach(uniqname => {
+                    const new_button = $(uniqname).clone();
+                    new_button.removeClass('uniqname-btn');
+                    new_button.addClass('uniqname-submit');
+                    new_button.attr('disabled', true);
+                    $('#modalAttendees').append(new_button);
+                });
+            })
+
+
```
Comment (bharadwaj-a) [inline attendance/index.html]: {nit} blank space

File: attendance/view.html [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+
+    {% include footer.html %}
+    {% include scripts.html %}
+    <script>
+        let monthIdxToString = {
+            0: "Jan", 1: "Feb", 2: "Mar", 3: "Apr",
+            4: "May", 5: "June", 6: "July", 7: "Aug",
+            8: "Sep", 9: "Oct", 10: "Nov", 11: "Dec"
+        }
+        let sortMeetings = (meetings) => {
```
Comment (bharadwaj-a) [inline attendance/view.html]: Again question of whether we are using ```const``` or ```let```

File: attendance/view.html [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+            })
+
+            fetch("http://backend.famnm.club/meeting")
+            .then( (res) => {
+                if (!res.ok) throw Error(res.statusText);
+                return res.json();
+            })
+            .then(sortMeetings)
+            .then(data => {
+
```
Comment (bharadwaj-a) [inline attendance/view.html]: {nit} awkward space

File: attendance/view.html [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+            .then( (res) => {
+                if (!res.ok) throw Error(res.statusText);
+                return res.json();
+            })
+            .then(sortMeetings)
+            .then(data => {
+                
+                data.forEach(meeting => {
+                    let meetDay = new Date(meeting.day);
+
```
Comment (bharadwaj-a) [inline attendance/view.html]: {nit} awkward space

File: attendance/view.html [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+                        if (!res.ok) throw Error(res.statusText);
+                        return res.json();
+                    }).then(attendees => {
+                        $(`#attendees-${meeting.meeting_id}`).text(attendees.map(a => a.uniqname).join(", "));
+                    });
+                })
+            })
+            .catch(console.log);
+
+            let populateUniqname = () => {
```
Comment (bharadwaj-a) [inline attendance/view.html]: ```let``` vs ```const```

File: attendance/view.html [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+
+            $('#hideUniqnameTable').click((e) => {
+                e.preventDefault();
+                $('#uniqname-meeting-header').hide();
+                $('#uniqname-meetings').hide();
+            });
+
+            $('#active-memb-button').click(() => {
+                $('#active-memb-list').empty();
+
```
Comment (bharadwaj-a) [inline attendance/view.html]: {nit} awkward space

File: attendance/view.html [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+                            <li>${row.uniqname}</li>
+                            `);
+                        });
+                    });
+                    $('#active-memb').show();
+                } else {
+                    $('#active-memb').hide();
+                }
+
+
```
Comment (bharadwaj-a) [inline attendance/view.html]: {nit} awkward space

File: attendance/view.html [hunk 18]
```diff
@@ -0,0 +1,205 @@
...
+                        });
+                    });
+                    $('#active-memb').show();
+                } else {
+                    $('#active-memb').hide();
+                }
+
+                
+            })
+
```
Comment (bharadwaj-a) [inline attendance/view.html]: {nit} awkward space

File: attendance/view.html [hunk 19]
```diff
@@ -0,0 +1,205 @@
...
+                <tr>
+                    <th scope="col">Meeting Date</th>
+                    <th scope="col">Type</th>
+                    <th scope="col">Description</th>
+                    <th scope="col">Attendees (uniqnames)</th>
+                </tr>
+            </thead>
+            <tbody id="meeting-body">
+
+            </tbody>
```
Comment (bharadwaj-a) [inline attendance/view.html]: {nit} combine into one ```<tbody ... />```

File: attendance/view.html [hunk 20]
```diff
@@ -0,0 +1,205 @@
...
+        <table class="table" id="uniqname-meetings">
+            <thead>
+                <tr>
+                    <th scope="col">Meeting Date</th>
+                    <th scope="col">Type</th>
+                </tr>
+            </thead>
+            <tbody id="uniqname-meeting-body">
+
+            </tbody>
```
Comment (bharadwaj-a) [inline attendance/view.html]: {nit} combine into one ```<tbody ... />```

File: attendance/view.html [hunk 21]
```diff
@@ -0,0 +1,205 @@
...
+    <h1 class="w-50 mx-auto">FAMNM Attendance Viewer</h1>
+
+    <div class='w-50 mx-auto d-flex justify-content-center'>
+        <a class="btn btn-success mx-3 my-3" id="csv-button" href="http://backend.famnm.club/export">Get Attendance Data as CSV</a>
+        <button class="btn btn-info mx-3 my-3" id="active-memb-button">Get Active Members</button>
+    </div>
+    <div class="mx-auto w-50 justify-content-center" id="active-memb">
+        <h2>Active FAMNM Members</h2>
+        <ol id="active-memb-list">
+        </ol>
```
Comment (bharadwaj-a) [inline attendance/view.html]: {nit} combine into one ```<ol ... />```

File: attendance/index.html (hunk 4 above)
Comment (eashwar) [reply attendance/index.html]: just tested; this change breaks bc `jquery` shenanigans.

File: attendance/index.html (hunk 7 above)
Comment (eashwar) [reply attendance/index.html]: i'll go with const

File: attendance/index.html (hunk 8 above)
Comment (eashwar) [reply attendance/index.html]: as with above, doing it like this breaks with jquery. keeping it as an empty thing with separate open and close tags does work.

**Pull request merged**